### DISCUSSION TITLE
SimCSE

### DISCUSSION THREAD

**Discussion by intrandom5** [boostcampaitech4nlp2/level1_semantictextsimilarity_nlp-level1-nlp-13#44] [Ideas]
논문에 대한 자세한 내용은 아래 링크를 참고. 본 discussion에서는 모델을 어떻게 우리 task에 적용할 것인지에 대한 관점에서만 말한다.
노션 정리 : https://www.notion.so/SimCSE-79699e63d59c431fa1158442b062c974
논문 원본 : https://arxiv.org/pdf/2104.08821.pdf

### 논문 프로세스
1. Unsupervised SimCSE
![image](https://user-images.githubusercontent.com/68216991/199171269-d1b25f7a-f9ef-499b-9e7c-14c69873a00d.png)
* transformer에는 last hidden state를 낼 때, dropout(p=0.1)을 적용한다.
* 이 dropout layer를 그대로 사용해 같은 문장에 대해 다른 feature를 뽑아낸다.
* 그러면 같은 문장임에도 서로 다른 feature가 만들어질테고, 이는 positive pair로 간주된다.
* 같은 원리로 뽑아진 같은 minibatch 내의 다른 문장들의 feature는 본 문장의 negative sample로 간주된다.
* 이를 통해서 두 문장의 유사성을 Unsupervised 방식으로 학습한다.
* 이는 cross-entropy objective with in-batch negatives를 통해 학습된다.
![image](https://user-images.githubusercontent.com/68216991/199171760-6b2716cc-5186-40b7-b434-5cc7a7a99357.png)
* 본 모델을 학습하는 데는 pre-train된 BERT, RoBERTa를 그대로 사용했고, dropout layer까지 그대로 사용했다.(추가적인 레이어 x) 위의 objective function을 통해 파라미터를 업데이트 하는 식으로 학습된다.

2. Supervised SimCSE
![image](https://user-images.githubusercontent.com/68216991/199172544-d9842a4f-453b-48da-ba2f-0ced31aef379.png)
* NLI task는 모델이 문장 임베딩을 학습하는데 가장 효율적인 방법이라는 것이 이미 이전 연구에서 입증됨.
* SimCSE는 이 NLI 데이터셋을 이용해 기존(문장 쌍에 대해 entailment, neutral, contradiction으로 분류하는 학습)과는 조금 다른 방법으로 학습을 진행.
* 논문에서 학습된 데이터셋은 한 문장에 사람이 직접 positive 문장과 negative 문장을 작성하는 식으로 구성되어 있기 때문에 하나의 문장 쌍에 positive, negative 문장 쌍이 모두 존재.
* 이를 이용해서 원래 objective function의 $(x_i,x_i^+)$를 $(x_i,x_i^+,x_i^-)로 확장해서 목적함수를 계산한다.
![image](https://user-images.githubusercontent.com/68216991/199173057-62ef66f1-4633-4a5d-9b0a-3dfb109bad5c.png)

3. Down-stream task에 적용.

### 우리 모델에 적용할 수 있는 방법
모델의 구조가 크게 추가되는 것은 없기 때문에, 구현은 할 수 있을 것 같다.
대회의 룰이 추가 데이터셋의 도움을 받으면 안되기 때문에, pretrain된 KoSimCSE를 쓴다거나, 다른 데이터셋으로 pre-training 하는 것은 안될 것 같다. 따라서 주어진 데이터셋만으로 아래의 과정을 수행.

\<Unsupervised SimCSE\>
1. Roberta-large pretrain 모델을 이용해 각 배치의 문장마다 문장 representation 쌍을 뽑아낸다. (pooler output 사용) (그 전에 같은 문장에 대해서 모델의 pooler output을 통해 나온 문장 embedding이 서로 다른지 확인하는 작업이 선행으로 필요.)

2. 해당 배치 내에서 위의 objective function을 계산하는 Loss를 구현(이 부분이 가장 어려울 것 같은데 github 코드를 참고해서 작성)

\<Supervised SimCSE\>
* 사실 이 부분은 우리 데이터셋이 같은 문장에 2개의 positive, negative 대응 쌍이 존재하지 않기 때문에 어려울 것 같은데 생각을 해봐야할 것 같다.

\<STS\>
* 위의 학습을 통해 fine-tuning된 RoBERTa 모델을 이용해 STS를 수행. 
* SBERT, BERT 둘 다 실험해 볼 필요가 있을 듯 하다.
